### PR TITLE
Issue/1010

### LINKED ISSUE #1010: Error reported by WP 5.2 error detection
### Reproduction Steps

Error reported by WP 5.2 error detection and sent via email

This was encountered on a LifterLMS 30-day demo trial site.

I don't know exactly how to recreate but wanted to record the error

### Error Messages / Logs

+ Include any relevant error messages or log files
```
Error Details
=============
An error of type E_ERROR was caused in line 69 of the file /srv/users/denim-butterfly/apps/denim-butterfly/public/wp-content/plugins/lifterlms/includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php. Error message: Uncaught Error: Call to a member function get_enrollment_status() on bool in /srv/users/denim-butterfly/apps/denim-butterfly/public/wp-content/plugins/lifterlms/includes/admin/post-types/meta-boxes/class.llms.meta.box.order.enrollment.php:69
Stack trace:
#0 /srv/users/denim-butterfly/apps/denim-butterfly/public/wp-admin/includes/template.php(1343): LLMS_Meta_Box_Order_Enrollment->output(Object(WP_Post), Array)
#1 /srv/users/denim-butterfly/apps/denim-butterfly/public/wp-admin/edit-form-advanced.php(670): do_meta_boxes(Object(WP_Screen), 'side', Object(WP_Post))
#2 /srv/users/denim-butterfly/apps/denim-butterfly/public/wp-admin/post.php(198): include('/srv/users/deni...')
#3 {main}
thrown 
```

### System Information

<details>
<summary>System Report</summary>
WP 5.3.2
LifterLMS 3.37.8
```
<!-- Paste your System Report below this line -->

```

</details>

### PR DESCRIPTION
**Pull request by eri-trabiccolo**
## Description
Fixes #1010 

@thomasplevy The reason I'm fixing this is that, despite what #1010 says, we got too many reports about this. They're not massive, but every now and then our support team has to deal with a customer reporting this, then asking _us_ what's that error etc. Let's fix this so that we can save time/resources :D. 

## How has this been tested?
Manually and with new unit tests.

## Screenshots 
n/a

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

**Comment by thomasplevy** [comment]
@eri-trabiccolo thanks for this, I reviewed briefly, I'll do some testing and full review later.

File: tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php [hunk 1]
```diff
@@ -77,4 +78,159 @@ public function test_save() {
...
+
+		$order_id   = $order->get( 'id' );
+		$product_id = $order->get( 'product_id' );
+		$student_id = $order->get( 'user_id' );
+
+		$order->set( 'anonymized', 'yes' );
+
+		ob_start();
+		$this->metabox->output();
+		$this->assertEquals( 'Cannot manage enrollment status for anonymized orders.', ob_get_clean() );
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php]: A test helper exists to change this to the suggestion

Source: https://github.com/gocodebox/lifterlms-tests/blob/a812e7be616331dc1a1acca2cba8a78b8dd04539/framework/traits/trait-llms-unit-test-assertions-output.php#L57-L71

```suggestion
		$this->assertOutputEquals( 'Cannot manage enrollment status for anonymized orders.', array( $this->metabox, 'output' ) );
```

File: tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php [hunk 2]
```diff
@@ -77,4 +78,159 @@ public function test_save() {
...
+
+		// Create a real order.
+		$order = $this->get_mock_order();
+
+		$order_id   = $order->get( 'id' );
+		$product_id = $order->get( 'product_id' );
+
+		$order->set( 'user_id', '' );
+
+		ob_start();
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php]: Like above, another helper: https://github.com/gocodebox/lifterlms-tests/blob/a812e7be616331dc1a1acca2cba8a78b8dd04539/framework/traits/trait-llms-unit-test-assertions-output.php#L42-L55

File: tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php [hunk 3]
```diff
@@ -77,4 +78,159 @@ public function test_save() {
...
+		// Create a real order.
+		$order = $this->get_mock_order();
+
+		$order_id   = $order->get( 'id' );
+		$product_id = $order->get( 'product_id' );
+		$student_id = $order->get( 'user_id' );
+
+		wp_delete_user( $student_id );
+
+		ob_start();
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php]: can use helper

File: tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php [hunk 4]
```diff
@@ -77,4 +78,159 @@ public function test_save() {
...
+	public function test_output_order_with_student() {
+
+		// Create a real order.
+		$order = $this->get_mock_order();
+
+		$order_id   = $order->get( 'id' );
+		$product_id = $order->get( 'product_id' );
+		$student_id = $order->get( 'user_id' );
+
+		ob_start();
```
Comment (thomasplevy) [inline tests/phpunit/unit-tests/admin/post-types/meta-boxes/class-llms-test-meta-box-order-enrollment.php]: https://github.com/gocodebox/lifterlms-tests/blob/a812e7be616331dc1a1acca2cba8a78b8dd04539/framework/traits/trait-llms-unit-test-assertions-output.php#L82

**Review by thomasplevy — CHANGES_REQUESTED** [review]
@eri-trabiccolo

I see one minor issue, see screencap:

Screenshot_2021-02-24_13-02-23

Pretty minor but should be easy enough to clean up really quick. I'm okay if we don't get into adding test cases for this particular thing but while we're looking at it that would also help make it very obvious that this student doesn't exist anymore instead of having a dead link

Secondly: my feedback is all about helper functions in the testing library. I don't actually care if you don't want to rewrite the tests to use the helpers but I wanted you to know they exist.

**Comment by eri-trabiccolo** [comment]
@thomasplevy done

**Comment by thomasplevy** [comment]
@eri-trabiccolo looks good, I noticed that the orders post table list has the same problem as the order metabox, I've just updated this locally instead of making you checkin on this again tomorrow. I'm going to merge into dev with my changes.

Thanks!

**Pull request merged**